### PR TITLE
Add stubs for `pyasn1`

### PR DESCRIPTION
**Pull request by Avasam**
For use with `ldap3` and `python-jose` stubs 😃 
Source: https://github.com/pyasn1/pyasn1/tree/db8f1a7930c6b5826357646746337dafc983f953/pyasn1

### CODE REVIEW

File: stubs/pyasn1/pyasn1/type/univ.pyi [hunk 1]
```diff
@@ -28,15 +33,16 @@ class Integer(base.SimpleAsn1Type):
     def __rmul__(self, value): ...
     def __mod__(self, value): ...
     def __rmod__(self, value): ...
-    def __pow__(self, value, modulo: Incomplete | None = ...): ...
+    # Accepts everything builtins.pow does
```
Comment (AlexWaygood) [inline stubs/pyasn1/pyasn1/type/univ.pyi:36]: 😱

File: stubs/pyasn1/pyasn1/type/univ.pyi [hunk 2]
```diff
@@ -247,21 +263,15 @@ class SequenceOfAndSetOfBase(base.ConstructedAsn1Type):
     def isInconsistent(self): ...
 
 class SequenceOf(SequenceOfAndSetOfBase):
-    __doc__: Incomplete
-    tagSet: Incomplete
-    componentType: Incomplete
-    subtypeSpec: Incomplete
-    typeId: Incomplete
+    __doc__: str
```
Comment (AlexWaygood) [inline stubs/pyasn1/pyasn1/type/univ.pyi:266]:
```suggestion
```

File: stubs/pyasn1/pyasn1/type/univ.pyi [hunk 3]
```diff
@@ -247,21 +263,15 @@ class SequenceOfAndSetOfBase(base.ConstructedAsn1Type):
...
+    typeId: int
 
 class SetOf(SequenceOfAndSetOfBase):
-    __doc__: Incomplete
-    tagSet: Incomplete
-    componentType: Incomplete
-    subtypeSpec: Incomplete
-    typeId: Incomplete
+    __doc__: str
```
Comment (AlexWaygood) [inline stubs/pyasn1/pyasn1/type/univ.pyi:270]:
```suggestion
```

File: stubs/pyasn1/pyasn1/type/useful.pyi [hunk 4]
```diff
@@ -1,31 +1,31 @@
...
     @property
     def asDateTime(self): ...
     @classmethod
     def fromDateTime(cls, dt): ...
 
 class GeneralizedTime(char.VisibleString, TimeMixIn):
-    __doc__: Incomplete
-    tagSet: Incomplete
-    typeId: Incomplete
+    __doc__: str
```
Comment (AlexWaygood) [inline stubs/pyasn1/pyasn1/type/useful.pyi:24]:
```suggestion
```

File: stubs/pyasn1/pyasn1/type/useful.pyi [hunk 5]
```diff
@@ -1,31 +1,31 @@
...
+    tagSet: TagSet
+    typeId: int
 
 class UTCTime(char.VisibleString, TimeMixIn):
-    __doc__: Incomplete
-    tagSet: Incomplete
-    typeId: Incomplete
+    __doc__: str
```
Comment (AlexWaygood) [inline stubs/pyasn1/pyasn1/type/useful.pyi:29]:
```suggestion
```

File: stubs/pyasn1/pyasn1/codec/ber/decoder.pyi [hunk 6]
```diff
@@ -1,325 +1,336 @@
...
 
-class AbstractDecoder:
-    protoComponent: Incomplete
+from pyasn1.type import base, char, univ, useful
+from pyasn1.type.base import Asn1Type
+from pyasn1.type.tag import TagSet
+
+_Unused: TypeAlias = object
+
+class AbstractDecoder(metaclass=ABCMeta):
```
[the remark this answers is not in the document]
Comment (Avasam) [reply stubs/pyasn1/pyasn1/codec/ber/decoder.pyi:12]: Ah if there's a significant runtime difference I'll remove it then. Good to know for mypy. I did notice for pyright

File: stubs/pyasn1/pyasn1/error.pyi [hunk 7]
```diff
@@ -1,11 +1,9 @@
-from _typeshed import Incomplete
-
 class PyAsn1Error(Exception): ...
 class ValueConstraintError(PyAsn1Error): ...
 class SubstrateUnderrunError(PyAsn1Error): ...
 
 class PyAsn1UnicodeError(PyAsn1Error, UnicodeError):
-    def __init__(self, message, unicode_error: Incomplete | None = ...) -> None: ...
+    def __init__(self, message, unicode_error: UnicodeError | object = ...) -> None: ...
```
[the remark this answers is not in the document]
Comment (Avasam) [reply stubs/pyasn1/pyasn1/error.pyi:6]: I think it's a mistake as well and meant None.

File: stubs/pyasn1/pyasn1/type/namedtype.pyi [hunk 8]
```diff
@@ -23,11 +21,11 @@ class NamedType:
     def getType(self): ...
 
 class OptionalNamedType(NamedType):
-    __doc__: Incomplete
+    __doc__: str
```
[the remark this answers is not in the document]
Comment (Avasam) [reply stubs/pyasn1/pyasn1/type/namedtype.pyi:24]: Just left it there from stubgen. Stubtest doesn't mind the removal

File: stubs/pyasn1/pyasn1/type/univ.pyi (hunk 1 above)
Comment (Avasam) [reply stubs/pyasn1/pyasn1/type/univ.pyi:36]: Luckily the Self return type means I can simplify all the overloads to `value: int | float | complex`. 😅

File: stubs/pyasn1/pyasn1/codec/cer/encoder.pyi [hunk 9]
```diff
@@ -8,12 +6,12 @@ class BooleanEncoder(encoder.IntegerEncoder):
...
-    DOT_CHAR: Incomplete
-    ZERO_CHAR: Incomplete
+    Z_CHAR: int
+    PLUS_CHAR: int
+    MINUS_CHAR: int
+    COMMA_CHAR: int
+    DOT_CHAR: int
+    ZERO_CHAR: int
     MIN_LENGTH: int
     MAX_LENGTH: int
```
[the remark this answers is not in the document]
Comment (Avasam) [reply stubs/pyasn1/pyasn1/codec/cer/encoder.pyi]: They make sense to me as constant ClassVar (and afaik you can't mix ClassVar and Final)

[the remark this answers is not in the document]
Comment (AlexWaygood) [reply stubs/pyasn1/pyasn1/codec/cer/encoder.pyi]: > They make sense to me as constant ClassVar (and afaik you can't mix ClassVar and Final)

Yup, `Final` in a stub implicitly indicates that it's a `ClassVar` :)

**Comment by AlexWaygood** [comment]
Looks like you removed a lot of `__eq__` and `__ne__` methods in b26e252 -- could you possibly add them back?

In _general_, we don't include methods or attributes in a subclass if they have the same signature as the same method or attribute as the superclass. But `__eq__` and `__ne__` methods are a bit special. Whether or not `__eq__` is overridden on a subclass has a big impact on mypy's very useful `--strict-equality` option, for example. (I don't know if pyright does similar analysis.) For example, you can see that my proposal to remove `decimal.Decimal.__eq__` in https://github.com/python/typeshed/pull/6877#issuecomment-1008350944 would have resulted in some false positives in `pandas` due to mypy's `--strict-equality` check.

In general, we need to be a lot more careful with dunder methods when it comes to removing them just because they're redundant overrides. E.g. https://github.com/python/mypy/issues/13800 was caused by removing a few `__hash__` methods in https://github.com/python/typeshed/pull/8465, and I'm not sure anybody's been able to figure out exactly _why_ yet!

File: stubs/pyasn1/pyasn1/codec/cer/encoder.pyi (hunk 9 above)
[the remark this answers is not in the document]
Comment (Avasam) [reply stubs/pyasn1/pyasn1/codec/cer/encoder.pyi]: Oops thought it was the other way around. Should I make them Final then?

**Comment by Avasam** [comment]
Oh 😯 No problem I'll add 'em back!

File: stubs/pyasn1/pyasn1/codec/cer/encoder.pyi (hunk 9 above)
[the remark this answers is not in the document]
Comment (AlexWaygood) [reply stubs/pyasn1/pyasn1/codec/cer/encoder.pyi]: If there's any chance that it's reasonable to override them in a subclass, they should be `ClassVar`s. If they're constants that should never be modified nor overridden in a subclass, they should be `Final`s. I haven't looked at the source code in _great_ detail, so I'll leave the decision to you :)

[the remark this answers is not in the document]
Comment (Avasam) [reply stubs/pyasn1/pyasn1/codec/cer/encoder.pyi]: I'm not 100% certain the intention behind the CHARs so I'll leave it like this.
LENGHTs are overridden below.

**Review by AlexWaygood — APPROVED** [review]
Thanks!

**Pull request merged**